### ISSUE TITLE
Mirrored strategy keras model.fit failure

### ISSUE THREAD

**Issue by plooney** [tensorflow/tensorflow#59091]
<details><summary>Click to expand!</summary> 
 
 ### Issue Type

Bug

### Have you reproduced the bug with TF nightly?

Yes

### Source

binary

### Tensorflow Version

2.12.0-dev20221202

### Custom Code

Yes

### OS Platform and Distribution

Linux 22.04

### Python version

Docker image

### CUDA/cuDNN version

From Docker image

### Current Behaviour?

```shell
Falls over with no error in model.fit using mirrored strategy with multiple gpus

I previously logged this error under issue.

 Crashes with exit code 135 in model.fit #58673 

 I have just gotten round to creating code to reproduce the bug.
```

### Standalone code to reproduce the issue

```shell
import tensorflow as tf
print("TensorFlow version:", tf.__version__)

mirrored_strategy = tf.distribute.MirroredStrategy()

with mirrored_strategy.scope():
  model = tf.keras.Sequential([tf.keras.layers.Dense(1, input_shape=(1,))])

model.compile(loss='mse', optimizer='sgd')

dataset = tf.data.Dataset.from_tensors(([1.], [1.])).repeat(100).batch(10)
model.fit(dataset, epochs=2)
print('Does it work?')
model.evaluate(dataset)
```

### Relevant log output

_No response_</details>

**Comment by plooney** [comment]
This works fine on my hardware with only using one device

`mirrored_strategy = tf.distribute.MirroredStrategy(devices=["/gpu:0"])`

**Comment by plooney** [comment]
It works up Tensorflow up to version 2.92

**Comment by sushreebarsa** [comment]
@plooney 
This issue seems to be Keras issue.  Please post this issue on [keras-team/keras repo.](https://github.com/keras-team/keras/issues) as Keras development is fully moving to [github.com/keras-team/keras](http://github.com/keras-team/keras). All issues and PRs related to keras will be addressed in that repo.
To know more see this TF forum discussion ; 
[https://discuss.tensorflow.org/t/keras-project-moved-to-new-repository-in-https-github-com-keras-team-keras/1999](https://discuss.tensorflow.org/t/keras-project-moved-to-new-repository-in-https-github-com-keras-team-keras/1999)
Thank you!

**Comment by plooney** [comment]
I think the issue may be in tensorflow itself as I cannot get a custom training loop to work either. For example

`import os
os.environ['TF_CPP_MIN_LOG_LEVEL'] = '0'
import tensorflow as tf
from tensorflow import keras
print("TensorFlow version:", tf.__version__)

mirrored_strategy = tf.distribute.MirroredStrategy()
global_batch_size = 10
dataset = tf.data.Dataset.from_tensors(([1.], [1.])).repeat(100).batch(global_batch_size)

with mirrored_strategy.scope():
  model = tf.keras.Sequential([tf.keras.layers.Dense(1, input_shape=(1,))])

optimizer = keras.optimizers.SGD(learning_rate=1e-3)
loss_fn = keras.losses.MeanSquaredError(reduction=tf.keras.losses.Reduction.NONE)

def compute_loss(labels, predictions):
    per_example_loss = loss_fn(labels, predictions)
    return tf.nn.compute_average_loss(per_example_loss, global_batch_size=global_batch_size)

def train_step(inputs):
    features, labels = inputs

    with tf.GradientTape() as tape:
        predictions = model(features, training=True)
        loss = compute_loss(labels, predictions)

    gradients = tape.gradient(loss, model.trainable_variables)
    optimizer.apply_gradients(zip(gradients, model.trainable_variables))
    return loss

@tf.function
def distributed_train_step(dist_inputs):
    per_replica_losses = mirrored_strategy.run(train_step, args=(dist_inputs,))
    return mirrored_strategy.reduce(tf.distribute.ReduceOp.SUM, per_replica_losses,
                                    axis=None)

dist_dataset = mirrored_strategy.experimental_distribute_dataset(dataset)

for dist_inputs in dist_dataset:
    print(distributed_train_step(dist_inputs))

print('Ran custom training')

#model.compile(loss='mse', optimizer='sgd')

#model.fit(dataset, epochs=2)
#print('Does it work?')
#model.evaluate(dataset)


`

**Comment by plooney** [comment]
@sushreebarsa could you repoen please

**Comment by sushreebarsa** [comment]
@plooney Thank you for the update!
Reopening this issue.

**Comment by jelalsultanov** [comment]
@plooney I had  same problem but was able to solve this with downgrading tensorflow from 2.11.0 to 2.9.2 as you commented. Thank you. My issue was that i have CNN classification model written with tf.keras and also I have 4 GPUs GeForce RTX 2070 11GB each, but initially my model was training on only single GPU, but my goal was to make it train on all 4GPUs. solution was just downgrading to 2.9.2

**Comment by ethanbonin** [comment]
Update on my part. 

I'm using `tensorflow==2.12.0`, and I am having the same issue. It crashes silently with exit code 135. Happens at the same spot everytime though.

It crashses at on the first epoch with these logs:

```bash
gradients/Reshape_grad/Shape:
gradients/Reshape_1_grad/Shape:
gradients/Reshape_2_grad/Shape:
gradients/Reshape_3_grad/Shape:
gradients/Reshape_4_grad/Shape:
gradients/Reshape_5_grad/Shape:
gradients/Reshape_6_grad/Shape:
```

```python
    def build_with_gpu(self, input_shape, X_train, y_train, X_test, y_test):
        print("----------------TRAINING WITH GPUs------------------")
        tf.debugging.set_log_device_placement(True)
        gpus = tf.config.list_physical_devices('GPU')
        for gpu in gpus:
            tf.config.experimental.set_memory_growth(gpu, True)
        logical_gpus = tf.config.list_logical_devices('GPU')
        strategy = tf.distribute.MirroredStrategy(logical_gpus)
        print(f'Number of devices: {strategy.num_replicas_in_sync}')

        BUFFER_SIZE = len(X_train)

        BATCH_SIZE_PER_REPLICA = self.batch_size
        GLOBAL_BATCH_SIZE = BATCH_SIZE_PER_REPLICA * strategy.num_replicas_in_sync

        train_dataset = tf.data.Dataset.from_tensor_slices((X_train, y_train)).shuffle(BUFFER_SIZE).batch(GLOBAL_BATCH_SIZE)
        test_dataset = tf.data.Dataset.from_tensor_slices((X_test, y_test)).batch(GLOBAL_BATCH_SIZE)

        train_dist_dataset = strategy.experimental_distribute_dataset(train_dataset)
        test_dist_dataset = strategy.experimental_distribute_dataset(test_dataset)
        steps_per_epoch = np.ceil(len(X_train) / GLOBAL_BATCH_SIZE)
        validation_steps = np.ceil(len(X_test) / GLOBAL_BATCH_SIZE)

        # A model, an optimizer, and a checkpoint must be created under `strategy.scope`.
        with strategy.scope():
            model = self.build_model(input_shape)
            model.compile(
                optimizer=self.optimizer,
                loss=self.loss,
                metrics=['accuracy', tf.keras.metrics.Precision()]
            )

        return (
            model,
            train_dist_dataset,
            test_dist_dataset,
            steps_per_epoch,
            validation_steps
        )
```

Then I call

```python
          model, train_dist_dataset, test_dist_dataset, steps_per_epoch, validation_steps = ml_model.build_with_gpu(
                input_shape=input_shape,
                X_train=X_train,
                y_train=y_train,
                X_test=X_test,
                y_test=y_test
            )
            history = model.fit(
                train_dist_dataset,
                validation_data=test_dist_dataset,
                epochs=ml_model.epochs,
                verbose=1,
                callbacks=[
                    tf.keras.callbacks.TensorBoard(log_dir=tensor_logs_dir),
                    # tf.keras.callbacks.ModelCheckpoint(filepath=checkpoint_prefix, save_weights_only=True),
                ],
                steps_per_epoch=steps_per_epoch,
                validation_steps=validation_steps
            )

```

**Comment by plooney** [comment]
@sushreebarsa Is there any update on this?

**Comment by plooney** [comment]
I've tried this with tf13 but the bug is still there. Is there going to be any resources allocated to trying to fix it?

**Comment by Venkat6871** [comment]
Hi,

Thank you for opening this issue. Since this issue has been open for a long time, the code/debug information for this issue may not be relevant with the current state of the code base.

The Tensorflow team is constantly improving the framework by fixing bugs and adding new features. We suggest you try the latest TensorFlow [version](https://pypi.org/project/tensorflow/) with the latest compatible hardware configuration which could potentially resolve the issue. If you are still facing the issue, please create a new GitHub issue with your latest findings, with all the debugging information which could help us investigate.

Please follow the [release notes](https://github.com/tensorflow/tensorflow/blob/master/RELEASE.md) to stay up to date with the latest developments which are happening in the Tensorflow space.
